tests/yast2_cmd/yast_firewall.pm

```diff
@@ -0,0 +1,91 @@
+# SUSE's openQA tests
+#
+# Copyright © 2019 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+#
+# All of cases is based on the reference:
+# https://documentation.suse.com/sles/15-SP1/single-html/SLES-admin/#id-1.3.3.6.13.6.13
+#
+# Tips: This testcase only runs on sles12 platforms.
+#
+# Summary: Setup SuSEfirewall2 and check if it works well
+# - enable firewall
+# - configure starup method interfaces, broadcast, services, and logging
+# - check the results
+# - remove the settings
+# - stop firewall
+# Maintainer: Shukui Liu <skliu@suse.com>
+
+use base 'consoletest';
+use strict;
+use warnings;
+use testapi;
+
+use utils;
+
+
+sub run {
+    select_console 'root-console';
+
+    #enable firewall
+    zypper_call('in yast2-firewall', exitcode => [0, 102, 103, 106]);
+
+    #change the startup setting and check it.
+    assert_script_run 'yast firewall startup atboot';
```

Comment:
I think that you can run all this test as a shell script, from L38 to L84 perhaps and only output on error

Reply:
See: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/8163/files#diff-0f05d6eb58e39b0b723d90d512a0813dR36

Reply:
the advantages of using the testapi(assert_script_run, validate_script_output) instead of shell script, are that
1) the check and verification process is automatically or smoothly done by the api for every statement.
2) the source code is more tidy compared to using awk/sed/grep for checking process.

Reply:
> the advantages of using the testapi(assert_script_run, validate_script_output) instead of shell script, are that
> 
>     1. the check and verification process is automatically or smoothly done by the api for every statement.

You can still do that, check: [this pr](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/7848/files#diff-d660c22784eea8d89f722ec81e49034bR44) and also [this job](https://openqa.suse.de/tests/3273709#external)

> 
>     2. the source code is more tidy compared to using awk/sed/grep for checking process.

You need to take into account that there might be [typing issues](https://progress.opensuse.org/issues/43889) which are a recurrying problem that can happen on any architecture where qemu is the backend. And right now if your code is merged, they "might" cause more trouble due to that issue for sle 15 SP2 and SLE 12 SP5.

So either unschedule them for these products, or follow the suggestions given in: https://mailman.suse.de/mailman/private/qa-maintenance/2019-May/013329.html

Reply:
I am trying to run all the test as a shell script(or maybe other languages such as python, perl or C in the future when needed), but I think we cannot avoid the typing issues entirely, because I will use the vulnerable API function(assert_script_run or type_string) at least once.